fw/cache.c

```diff
@@ -3108,99 +3106,163 @@ tfw_cache_can_use_stale(TfwHttpReq *req, TfwCacheEntry *ce, long age)
+		tfw_http_conn_msg_free((TfwHttpMsg *)req);
+		return;
 	}
 
+	action((TfwHttpMsg *)req);
+}
+
+static void
+cache_do_service_request(TfwHttpReq *req, tfw_http_cache_cb_t action,
+			 TfwCacheEntry *ce, long age)
+{
+	TfwHttpResp *resp = NULL;
+	unsigned int id = 0;
+
 	T_DBG("Cache: service request [%p] w/ key=%lx, ce=%p", req,
 	      ce->trec.key, ce);
 
 	TFW_INC_STAT_BH(cache.hits);
 
 	if (!tfw_handle_validation_req(req, ce))
-		goto put;
+		return;
 
 	/*
 	 * If the stream for HTTP/2-request is already closed (due to some
 	 * error or just reset from the client side), there is no sense to
-	 * forward request to the server.
+	 * forward response to the client.
 	 */
 	if (TFW_MSG_H2(req)) {
 		id = tfw_h2_req_stream_id(req);
 		if (unlikely(!id)) {
 			tfw_http_conn_msg_free((TfwHttpMsg *)req);
-			goto put;
+			return;
 		}
 	}
 
-	if (!stale)
-		resp = tfw_cache_build_resp(req, ce, age);
+	resp = tfw_cache_build_resp(req, ce, age);
+	if (unlikely(!resp)) {
+		tfw_http_conn_msg_free((TfwHttpMsg *)req);
+		T_ERR("Cache: Can't allocate response.");
+		return;
+	}
 
 	/*
 	 * The stream of HTTP/2-request should be closed here since we have
 	 * successfully created the resulting response from cache and will
 	 * send this response to the client (without forwarding request to
 	 * the backend), thus the stream will be finished.
-	 *
-	 * The stream can be unlinked only for non stale responses. If response
-	 * is stale, request will be forwarded to server, some forwardning
-	 * functions requires alive stream. E.g: @tfw_http_req_evict_dropped().
 	 */
-	if (resp && TFW_MSG_H2(req)) {
+	if (TFW_MSG_H2(req)) {
 		id = tfw_h2_req_stream_id(req);
 		if (unlikely(!id)) {
 			tfw_http_msg_free((TfwHttpMsg *)resp);
 			tfw_http_conn_msg_free((TfwHttpMsg *)req);
-			goto put;
+			return;
 		}
 		tfw_h2_req_unlink_stream(req);
 	}
-out:
-	if (!stale && !resp && (req->cache_ctl.flags & TFW_HTTP_CC_OIFCACHED)) {
+
+	action((TfwHttpMsg *)req);
+}
+
+/*
+ * Try to service request from the cache. If it allowed use stale cached
+ * response.
+ *
+ * NOTE: This function is responsible for releasing non stale cache entry
+ * using @tdb_rec_put(). Stale cache entry must be assigned to request then
+ * released during response forwarding or request destruction.
+ */
+static bool
+cache_try_service_request(TfwHttpReq *req, tfw_http_cache_cb_t action)
+{
+	TDB *db = node_db();
+	TdbIter iter;
+	TfwCacheEntry *ce = NULL;
+	long lifetime;
+	long age;
+
+	if (!(ce = tfw_cache_dbce_get(db, &iter, req))) {
+		T_DBG3("%s: db=[%p] req=[%p] CE has not been found\n",
+		       __func__, db, req);
+		return false;
+	}
+	__tfw_dbg_dump_ce(ce);
+
+	age = tfw_cache_entry_age(ce);
+
+	if (!(lifetime = tfw_cache_entry_is_live(req, ce, age))) {
+		bool success;
+
+		T_DBG3("%s: db=[%p] req=[%p] ce=[%p] CE is not alive\n",
+		       __func__, db, req, ce);
+
+		if ((success = tfw_cache_can_use_stale(req, ce, age)))
+			cache_do_service_request_stale(req, action, ce, age);
```

Comment:
Who drop `ce` after `cache_do_service_request_stale`?

Reply:
On next line there is a comment. It will be freed during stale forwarding or in req destructor.